### PR TITLE
feat: better prng

### LINKED ISSUE #879: Merge core/random and x/rand
I would like to merge x/rand with core/random, but these two packages doesn't seem to share the same underlying implementation of PRNG.

I would also like to see more support for random in core, e.g.

1. Generate `Int/Int64/Float/Double` with range.
2. More random operation on array:
   1. `sample[T](Array[T], ~k: Int) -> Array[T]`
   2. `choice[T](Array[T]) -> T`

### PR DESCRIPTION
**Pull request by qazxcdswe123**
This PR introduced new PRNG, with output exactly the same as go's `math/rand/v2`.

It is also cryptographically secure, but comes with a more strict interface for seeding.

close #879

### CODE REVIEW

File: random/rand.mbt [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+/// Currently we only support [chacha8] as the source of randomness.
+struct Rand {
+  src : ChaCha8
+}
+
+/// Create a new random number generator with [seed].
+/// @alert unsafe "Panic if seed is not 32 bytes long"
+pub fn Rand::new(
+  ~seed : Bytes = string_to_bytes("ABCDEFGHIJKLMNOPQRSTUVWXYZ123456")
```
Comment (qazxcdswe123) [inline random/rand.mbt:23]: Go's rand is automatically seeded by its runtime, which is not supported by moonbit yet. This interface is somewhat awkward.

File: random/rand.mbti [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package moonbitlang/x/rand
+
+// Values
+
+// Types and methods
+type ChaCha8
+
+type Rand
```
Comment (bobzhang) [inline random/rand.mbti:8]: please remove this file

File: random/random.mbti [hunk 3]
```diff
@@ -3,13 +3,21 @@ package moonbitlang/core/random
...
 
 // Types and methods
-type RandomState
-impl RandomState {
-  gen_double(Self) -> Double
-  gen_int(Self) -> Int
-  gen_int64(Self) -> Int64
-  init_state(~seed : Int = ..) -> Self
-  new() -> Self
+type ChaCha8
```
Comment (bobzhang) [inline random/random.mbti:6]: why this is exposed

Comment (qazxcdswe123) [reply random/random.mbti:6]: It was removed in x , but out of sync here. Sorry

File: random/rand_source_chacha.mbt [hunk 4]
```diff
@@ -0,0 +1,469 @@
+// Copyright 2024 International Digital Economy Academy
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (bobzhang) [inline random/rand_source_chacha.mbt:3]: is it easy to move it into random/internal/source_chacha?

File: random/rand.mbt [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+  src : ChaCha8
+}
+
+/// Create a new random number generator with [seed].
+/// @alert unsafe "Panic if seed is not 32 bytes long"
+pub fn Rand::new(~seed : Bytes = b"ABCDEFGHIJKLMNOPQRSTUVWXYZ123456") -> Rand {
+  if seed.length() != 32 {
+    abort("seed must be 32 bytes long")
+  }
+  let src = ChaCha8::new(seed)
```
Comment (bobzhang) [inline random/rand.mbt:26]: no need panic? you can always make it length 32,right?

File: random/rand.mbt [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+  assert_eq!(n, exp)
+}
+
+/// [int] Return a non-negative pseudo-random 31-bit integer as an Int.
+pub fn Rand::int(self : Rand) -> Int {
+  self.next().lsr(33).to_int()
+}
+
+/// [int64] returns a non-negative pseudo-random 63-bit integer as an Int64
+pub fn Rand::int64(self : Rand) -> Int64 {
```
Comment (bobzhang) [inline random/rand.mbt:53]: can you document its range precisely?

File: random/rand.mbt [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+test "next" {
+  let r = Rand::new()
+  let n = r.next()
+  let exp = 13219109469176600229UL
+  assert_eq!(n, exp)
+}
+
+/// [int] Return a non-negative pseudo-random 31-bit integer as an Int.
```
Comment (bobzhang) [inline random/rand.mbt:47]: precise range

File: random/random.mbti [hunk 8]
```diff
@@ -3,13 +3,19 @@ package moonbitlang/core/random
...
-type RandomState
-impl RandomState {
-  gen_double(Self) -> Double
-  gen_int(Self) -> Int
-  gen_int64(Self) -> Int64
-  init_state(~seed : Int = ..) -> Self
-  new() -> Self
+type Rand
+impl Rand {
+  double64(Self) -> Double
```
Comment (bobzhang) [inline random/random.mbti:8]: just double?

File: random/random.mbti [hunk 9]
```diff
@@ -3,13 +3,19 @@ package moonbitlang/core/random
...
+  int(Self) -> Int
+  int64(Self) -> Int64
+  int64N(Self, Int64) -> Int64
+  intN(Self, Int) -> Int
+  new(~seed : Bytes = ..) -> Self
```
Comment (bobzhang) [inline random/random.mbti:13]: what does `intN` `N` mean here, can you give it a more meaningful name?

File: random/rand.mbt (hunk 5 above)
Comment (qazxcdswe123) [reply random/rand.mbt:26]: Yes it is possible, but another benefits for using chacha as random source is that it is cryptographically secure.

If we just pad it to 32, it no longer hold this property.

File: random/random.mbti (hunk 8 above)
Comment (qazxcdswe123) [reply random/random.mbti:8]: I tried to add float but float show is not implemented and thus cannot be tested, so I git stashed it and forgot..

File: random/random.mbti (hunk 9 above)
Comment (qazxcdswe123) [reply random/random.mbti:13]: `N` means range, thats what go's API looks like.

Probably rename to `int_range`?

File: random/rand_source_chacha.mbt (hunk 4 above)
Comment (qazxcdswe123) [reply random/rand_source_chacha.mbt]: Done

File: random/random.mbti [hunk 10]
```diff
@@ -3,13 +3,20 @@ package moonbitlang/core/random
...
-  init_state(~seed : Int = ..) -> Self
-  new() -> Self
+type Rand
+impl Rand {
+  double64(Self) -> Double
+  float32(Self) -> Float
+  int(Self) -> Int
+  int64(Self) -> Int64
+  int64N(Self, Int64) -> Int64
+  intN(Self, Int) -> Int
```
Comment (tonyfettes) [inline random/random.mbti:13]: Use optional argument for `intN`? for example `int(Self, ~limit?:Int)`

**Review by tonyfettes — CHANGES_REQUESTED** [review]

File: random/random.mbti (hunk 10 above)
Comment (bobzhang) [reply random/random.mbti:13]: uuml avoid allocation -- imperative API

**Review by tonyfettes — APPROVED** [review]
LGMT

**Pull request merged**